### DISCUSSION TITLE
Development Thread About CTRLR 5.6 for Apple Silicon M1/M2 [ARM64]

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#565] [General]
ARM64 being the new standard architecture for the latest Apple computers with their Apple Silicon M1 & M2 CPUs, we decided to open a thread about the current status of CTRLR for Apple M1/M2 setups.
More to come soon.

Damien

**Comment by damiensellier** [comment]
NOTE : version deprecated v5.6.31b is now available : 

https://github.com/RomanKubiak/ctrlr/discussions/700#discussion-7901494

___

There's a macOS M1 version for Ctrlr 5.6 (commit f452347) build available, I downloaded it when I saw it but I don't have a M1 or M2. Here are the files :

Ctrlr 5.6 Mac ARM Standalone: 
[Ctrlr-5.6-f452347-App.zip](https://github.com/RomanKubiak/ctrlr/files/10947199/Ctrlr-5.6-f452347-App.zip)

Ctrlr 5.6 Mac ARM Appex: 
[Ctrlr-5.6-f452347-appex.zip](https://github.com/RomanKubiak/ctrlr/files/10947200/Ctrlr-5.6-f452347-appex.zip)

Ctrlr 5.6 Mac ARM AU: 
[Ctrlr-5.6-f452347-component.zip](https://github.com/RomanKubiak/ctrlr/files/10947201/Ctrlr-5.6-f452347-component.zip)

Ctrlr 5.6 Mac ARM VST2: 
[Ctrlr-5.6-f452347-vst.zip](https://github.com/RomanKubiak/ctrlr/files/10947202/Ctrlr-5.6-f452347-vst.zip)

Ctrlr 5.6 Mac ARM VST3: 
[Ctrlr-5.6-f452347-vst3.zip](https://github.com/RomanKubiak/ctrlr/files/10947203/Ctrlr-5.6-f452347-vst3.zip)

It seems that the midi I/O are not working properly on this version. 
The latest intel version of CTRLR is 5.6.1 based on JUCE 6.0.7.
The first thing to do would be to update CTRLR to work with the latest JUCE version 7 which is fully compatible with Apple Silicon computers.

**Reply by bijlevel** [reply]
The link to the commit f452347 tells me: "This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository". Do I actual need this commit or are the zip files good enough for compiling?

**Reply by damiensellier** [reply]
This is the zip corresponding : 
https://github.com/RomanKubiak/ctrlr/archive/f452347540725588736a425f38c67c3b514ef24d.zip

I'm not sure the uodated files have been merged to the main repo, Please check what @jmechnich changed to replicate in the latest version.

**Reply by bijlevel** [reply]
Thanks. It's what dobo365 wrote, interested, but not yet a M1/M2 computer. But each day :-) I'm tempted to buy one, but not sure which one is the best for me. And if I wait, maybe Apple has yet another better one...

**Reply by jmechnich** [reply]
Hi and sorry for the late reply,
the branch with my changes from back in the day still exists here: https://github.com/jmechnich/ctrlr/tree/fix-au-type

It looks like the merge to the parent repo has been reverted at some point for reasons unknown to me. As I have not continued using CTRLR since then, unfortunately I am unable to tell you how the state of affairs is with current versions of macOS.

My takeaway was that a major rewrite to the software is necessary to have things working smoothly again.

**Reply by damiensellier** [reply]
Hi Jörg, thanks for the info, your version always worked well for me. I never used it on the newer macOS though. Have a nice weekend!

**Reply by bijlevel** [reply]
Today I started using my new Mac Studio M2 Ultra and I could start CTRLR 5.6.23 successfully after downloading Rosetta. The only thing I had to do was reconnecting my resources (Knobman sliders) again and after that things seem the same as on an intel Mac. So maybe no need for an ARM version, though I'm interested and willing to compile one if I have the proper resources...

**Reply by bijlevel** [reply]
I had a better look at this thread ;-) and downloaded the zip file for the standalone version, supposedly running on M1/M2 Mac. (It seems this is version 5.6.23, which was the result of my compile efforts some time ago.) I can't check if it really runs natively (anyone knows how to?), because to run the previous (none M1 version) I had to download Rosetta and after that it did run fine. But maybe this so called M1-version does also use Rosetta. I don't know how to disable Rosetta for this particular version... (But no problems with MIDI with this version.)

**Reply by damiensellier** [reply]
Hi, here’s how you can check if an app is running on Rosetta or M1 native architecture:

1. Bring up the Spotlight search by hitting Command + Space on your M1 Mac.
2. Type in “Activity Monitor,” and hit return.
3. You’ll see a dialog box with parameters of different apps. Click on the CPU section in the top bar.
4. Once the section loads up, you’ll see a column named “Kind”. If the app says “Intel” under that, you should download the native version if available.

Let us know if it runs natively ;) thanks

**Reply by bijlevel** [reply]
It says Apple! Thanks Damian for pointing out the way to check if an app is using Rosetta. It did show me some Intel apps, so very useful! As I wrote earlier, my findings with this ARM version are that it all works as expected, also MIDI.

**Reply by bijlevel** [reply]
I've been working now for a while with the ARM64 version and it behaves rather well. One thing though is rather annoying. Every time when I want to open a panel from a certain folder on my computer it asks me to permit acces to that folder. Sometimes the dialog window, asking this permission, appears more than 6 times or so!

**Reply by jmechnich** [reply]
@bijlevel you could try removing the permissions for Ctrlr completely and see if that helps:

1. Open **System Settings**
2. Select **Privacy & Security** from the Navigation Bar
3. Select **Files and Folders** from the **Privacy** section
4. Select **Ctrlr** and press the '-' sign at the very bottom of the window to remove it from the list

After re-opening Ctrlr, macOS will ask you again for permission to access certain folders and, fingers crossed, it should now be stored correctly. It seems to work on my end (on macOS 14.2).

I have also built a new binary using the current Xcode that I could release after a bit more testing.

**Reply by bijlevel** [reply]
Thanks for the help, but removing CTRLR form the list made it worse. I now had to give permission for every file/folder on my system. I had to force quit CTRLR to stop it asking. After that I reset in System Settings for CTRLR the "old" permissions. See screenshot. Still CTRLR asks for access, but now at most "only" 6 times or so...

<img width="500" alt="Screenshot 2023-12-17 at 20 12 35" src="https://github.com/RomanKubiak/ctrlr/assets/11859363/f866e650-57fd-4917-bff1-7484014e9cda">

**Reply by bijlevel** [reply]
I'm very curious to your new build! If it works and solves these permission issues (and maybe some more lurking issues) that would be wonderful! I planned myself also to do a build on my new Mac, if I can find the time. If I start, do I need to do some specific things? Of course I have install Xcode, but is there more, like using the latest Juce etc? I can use all the help, or maybe we can even help each other?

Edit: I'm on Ventura 13.6.1

**Reply by damiensellier** [reply]
same here, looking forward to your new build and a bit of explanation on how to export it from xcode :-) Thanks guys for making this possible

**Reply by midibox** [reply]
Some weeks ago I created a build for Apple Silicon with a small modification to get my LUA scripts working again.
It can be found here: http://www.ucapps.de/midibox_sid/ctrlr/Ctrlr__tk_changes_for_midibox__20231026.zip
For some reasons, I'm not affected by the permission issue...

Background story: https://github.com/RomanKubiak/ctrlr/issues/560#issuecomment-1771700555

Best Regards, Thorsten.

**Reply by damiensellier** [reply]
Hi Thorsten, I just realised you're the person behind the midibox project. I was dreaming about building a midibox LC console and a midibox-SID when I was a kid. Glad to have you here and thanks for your input on the apple silicon build!
Damien

**Reply by bijlevel** [reply]
Thanks a lot for this build! The permissions issue seems indeed gone.

**Reply by bijlevel** [reply]
I thought the permission issue was gone, but it's still there. Very annoying!

**Reply by bijlevel** [reply]
I did manage to build CTRLR as an app on my new Mac Studio (Xcode 15.1, Juce 6.1.4) and the permission issue is now gone. From the messages in Xcode I understand there is still a problem with the build of the AU plugin. I hope to overcome it, because having a AU plugin would be really nice! Also it seams rather easy to build for silicon mac. It is an option when choosing how to build.

**Reply by damiensellier** [reply]
@bijlevel That's a great news! 
Can you please share your ctrlr.jucer file here?
Thank you

**Reply by bijlevel** [reply]
I did downgrade to Xcode v14.3.1 because of the problems with building an AU. The AU build did succeed with no errors, but when I wanted to load the plug-in in my DAW it failed examination. So that needs further investigation. The CTRLR app though works fine. The build defaults now to Silicon Mac, but that's OK with me. I have added the Projucer file to this post. I did use Juce v 6.1.4. Make sure the global path is set correctly to point to this version.

[Ctrlr.jucer.zip](https://github.com/RomanKubiak/ctrlr/files/13754343/Ctrlr.jucer.zip)

**Reply by dobo365** [reply]
@bijlevel Super good news! Thx a lot for this contribution.
I'm now planning to purchase a second hand Mac Mini M1 as the prices are reasonable.
The only goal would be to generate M1 versions of my panels. Standalone, VST3 and AU.
The Standalone is great to have as requested by many people BUT the key is the VST3 because on M1, Cubase 12 is only supporting a native VST3. AU is also nice to have for the Logic users of course.
I don't know if a native AU is supported by Cubase 12 on M1.

**Reply by bijlevel** [reply]
So far only the CTRLR app is running natively OK. AU plugin does not (yet) work. It loads OK in my DAW (Digital Performer, DP) after examination by DP, but when I try to use the CTRLR AU plugin, I get a serious error and have to restart DP. I did no yet build a VST3, but DP is known for having many issues with VST3. I will try these plugins in another DAW this weekend.

**Reply by bijlevel** [reply]
Good news: I did load the CTRLR AU plugin in 2 other programs that recognize AU's and in both programs the CTRLR AU did load fine. So it is probably a problem of Digital Performer. If you want the app and the plugin (build for Silicon Mac) I can upload them. Later on I will build the VST3 and hopefully that one does also load! Be aware that System Settings/Privacy & Security/ will complain, but it is easy to allow MacOS to open the app (and plugin) using Privacy & Security settings.

**Reply by bijlevel** [reply]
Today I did build the VST3 plugin and I'm happy to tell that the plugin does load in both Mixbus and Digital Performer. The fact that it loads (and works) in DP is very promising, because DP used to be very picky about VST3! I'm not sure though if the generated  plugins are native Mac. I don't know how to check this. The app is native for sure!

**Reply by bijlevel** [reply]
I started with implementing a mouseUp event for a slider. I did copy all the mouseDown code I could find and copied it into to the various files and changed every mouseDown text with mouseUp text. Sometimes Xcode gave an error after pasting, so that code I did remove. After I had copied/pasted many, many code from many places in the source and Xcode didn't show errors, I did a build. It build without errors, but the interface showed this:

<img width="841" alt="Screenshot 2023-12-24 at 19 08 09" src="https://github.com/RomanKubiak/ctrlr/assets/11859363/9e24fc7c-3327-4e6d-9a89-83dbd9d4f964">

So it seems it needs some more work in the UI... I'm on my way, but not yet reached the finishline! 😉 If someone has tips about how to proceed, I would be grateful. In the meantime I will keep on trying.

**Reply by damiensellier** [reply]
Did you copy paste everything from both the .cpp and the headers .h ?  

You should also take a look at CtrlrComponent.cpp on line 103 it refers to the different panel properties for the generic components : 
https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/UIComponents/CtrlrComponents/CtrlrComponent.cpp#L103

A new line defining the mouseUp event must be added : 

`setProperty (Ids::componentLuaMouseUp, COMBO_ITEM_NONE);`


then define the MouseUp function on line 186, always in CtrlrComponent.cpp : 


```
void CtrlrComponent::mouseUp(const MouseEvent &e)
{
    if (mouseUpCbk && !mouseUpCbk.wasObjectDeleted())
	{
		if (mouseUpCbk->isValid())
		{
			owner.getOwnerPanel().getCtrlrLuaManager().getMethodManager().call (mouseUpCbk, this, e);
		}
	}
}

```

then on line 448 must be added the related callback function on this particular event (after the double click definition) : 

https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/UIComponents/CtrlrComponents/CtrlrComponent.cpp#L441
```

	else if (property == Ids::componentLuaMouseUp)
	{
		if (isInvalidMethodName (getProperty(property)))
			return;

		mouseUpCbk = owner.getOwnerPanel().getCtrlrLuaManager().getMethodManager().getMethod(getProperty(property));
	}
```

And also on the related header [CtrlrComponent.h](https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/UIComponents/CtrlrComponents/CtrlrComponent.h ) , you should add on line 92 : 
`        virtual void mouseUp(const MouseEvent &e);`

Then on line 132 you should add the mouseUpCbk variable : 

`WeakReference <CtrlrLuaMethod> mouseMoveCbk, mouseDownCbk, mouseDragCbk, mouseDoubleClickCbk, mouseUpCbk ;
`


I think this would work with all that. This should add the mouseUp event callback dropdown for every "generic component" like sliders and buttons. 

Hope it helps ;)

Edit: I added the mouseUp function. The different sliders cpp require the mouseUp function to be modified too.

**Reply by bijlevel** [reply]
HI Damien, I did double check, but I did add indeed the needed mouseUp code to all those places you indicate, and I think even to some other places. The fact that the build succeeds without errors means (I hope) this part is OK. Still the mouseUp callback has to appear in the panel interface...

**Reply by nutsynick** [reply]
i have an m4 mac and the standalown works but none of the others show in ableton 12

**Reply by bijlevel** [reply]
Did you have to move the standalone out of quarantine? If so then you should do the same with the plugins.

**Comment by dobo365** [comment]
Also interested in that one and in the same situation as you; no M1/2 computer ;-)
Thx for creating this summary thread!

**Reply by damiensellier** [reply]
I know John @goodweather worked on the PC version of CTRLR with JUCE 7.3 in April but not sure if he succeeded with that. 
If all the files are working well together with the latest LUA, LUA Bind, JUCE etc that would be a good start to upgrade both PC, Intel Mac and M1/M2 Ctrlr versions. I'll ask him via email so that we can use his github fork as a starter.

**Comment by nutsynick** [comment]
Tried that nothing happened with vst2 and vst3--Sent with mail.com Mail appOn 13/02/2025, 19:23 bijlevel ***@***.***> wrote:

Did you have to move the standalone out of quarantine? If so then you should do the same with the plugins.

**Reply by bijlevel** [reply]
What about the AU plugin?

**Reply by damiensellier** [reply]
Hi, we covered the problem with Ableton Live. There's a problem with codesigning.
This is a link I posted in the ctrlrx 5.6.31b release discussion:

https://github.com/user-attachments/files/18652941/CtrlrX.5.6.31.macOS.Intel.VST.NoCodesignForLive.zip

This build does not have the local signature so it's not blacklisted by live when scanning for new vst plugins.

Hope it helps. I'll post the link to this specific issue below 👇 

Damien

**Reply by damiensellier** [reply]
Ctrlrx v5.6.31b fixed the vst not showing up in live ableton 
https://github.com/RomanKubiak/ctrlr/discussions/700#discussion-7901494

But, vst2 has codesign issue so here is the fix : 
https://github.com/damiensellier/CtrlrX/issues/59

There's a command line to uncodesign the binary : 
https://github.com/damiensellier/CtrlrX/issues/47#issuecomment-2643746885

`codesign -f -s - your_path_to_CtrlrX.vst_or _your exported_instances`

It's going to be required for all exported instances.

Damien

**Comment by nutsynick** [comment]
That's in there now but still not seen by Ableton --Sent with mail.com Mail appOn 13/02/2025, 19:31 bijlevel ***@***.***> wrote:

What about the AU plugin?

**Reply by bijlevel** [reply]
I will look into it. On my side it is no problem, because I did build the app and plugins ;-), so it is trusted by my Mac. I will  download the builds, which forces my Mac to put them in quaratine, and will see what happens. But not tonight, maybe tomorrow...

**Comment by nutsynick** [comment]
Awesome...IAM a beginner at this code stuff.....but used ctrl for years on an intel Mac, I luv it--Sent with mail.com Mail appOn 13/02/2025, 19:37 bijlevel ***@***.***> wrote:

I will look into it. On my side it is no problem, because I did build the app and plugins ;-), so it is trusted by my Mac. I will  download the builds, which forces my Mac to put them in quaratine, and will see what happens. But not tonight, maybe tomorrow...

**Reply by bijlevel** [reply]
You should know that the Intel version works like a charm on a Silicon Mac, but of course the Mac uses Rosetta to accomplish this. Speed is not an issue, I noticed. BTW I never found it a problem to use (old) Intel stuff this way on my Silicon Mac. Maybe you can try that (also).

**Reply by bijlevel** [reply]
I noticed you're probably using not the latest downloads. These can be found here: #700 So maybe you try them first. Also you should be aware of the problems with Ableton as indicated by @damiensellier. I'm not sure if the issue is already solved. AFAIK he is working on a CTRLR update (5.6.32)

**Comment by nutsynick** [comment]
Many many thanks guys, I'll keep u posted --Sent with mail.com Mail appOn 13/02/2025, 21:20 Damien Sellier ***@***.***> wrote:

Hi, we covered the problem with Ableton Live. There's a problem with codesigning.
This is a link I posted in the ctrlrx 5.6.31b release discussion:
https://github.com/user-attachments/files/18652941/CtrlrX.5.6.31.macOS.Intel.VST.NoCodesignForLive.zip
This build does not have the local signature so it's not blacklisted by live when scanning for new vst3 plugins.
Hope it helps. I'll post the link to this specific issue below 👇
Damien

**Comment by damiensellier** [comment]
For information, I'll soon release v5.6.32 which fixes a few issues mentioned here. 
Probably early March.

If you have any issues, don't hesitate to report your bug with details.

Thanks 

Damien

**Comment by dnaldoog** [comment]
Hey Damien,

Got a user om M1 for the Alpha Control panel asking about this message:

![{F645F061-FD17-47AA-9EEB-B73E866064AB}](https://github.com/user-attachments/assets/68238a78-eef9-4f5c-a668-0505451fa478)

I found this code online and advised him to modify to suit - don't know if he actually did it or whether it's correct.

>It looks like you may have to do something like running this command in a terminal:
`codesign -f -s - 'target/bundled/plugin-name.vst3`
>I guess wherever the plugin Alpha Control.vst3 is saved to.

**Reply by damiensellier** [reply]
Whitelisting command :
https://github.com/damiensellier/CtrlrX/wiki/Whitelisting-Ctrlr-on-macOS-and-avoid-quarantine

Codesigning for local use:
https://github.com/damiensellier/CtrlrX/wiki/Replacing-CtrlrX-exported-instance-corrupted-codesigning-on-macOS

**Comment by damiensellier** [comment]
Ctrlrx is now fully working for Mac Silicon CPUs natively : https://github.com/damiensellier/CtrlrX/releases/tag/5.6.33

**Comment by dnaldoog** [comment]
Wow thank you - I've passed the info on.
